packages/liveblocks-core/src/lib/AsyncCache.ts

```diff
@@ -0,0 +1,345 @@
+import type { Callback, Observable, UnsubscribeCallback } from "./EventSource";
+import { makeEventSource } from "./EventSource";
+
+const DEDUPLICATION_INTERVAL = 2000;
+
+type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] };
+
+type AsyncFunction<T, A extends any[] = any[]> = (...args: A) => Promise<T>;
+
+type AsyncCacheOptions = {
+  deduplicationInterval?: number;
+};
+
+type AsyncCacheItemOptions = WithRequired<
+  AsyncCacheOptions,
+  "deduplicationInterval"
+>;
+
+type InvalidateOptions<TData = any> =
+  | { clearData?: false; setOptimisticData?: never }
```

Comment:
When do you think the option `clearData` will be used?

Reply:
For use-cases that should work in a stale-while-revalidate fashion instead of the default which is to never return stale data.

The use-case's own `get()` function would check if the key needs to be invalidated (e.g. based on time), then do `invalidate({ clearData: false })`, call the cache's `get()` function but without awaiting it (just to schedule it) and immediately return the cached/stale data.